metricflow/model/semantics/join_validator.py

```diff
@@ -18,6 +18,23 @@ class DataSourceIdentifierJoinType:
     right_identifier_type: IdentifierType
 
 
+@dataclass(frozen=True)
+class DataSourceIdentifierJoin:
+    """Link between an identifier and a DataSourceIdentifierJoinType."""
+
+    identifier_reference: IdentifierReference
+    join_type: DataSourceIdentifierJoinType
+
+
+@dataclass(frozen=True)
+class DataSourceJoin:
+    """Links between two data sources and the valid way to join them."""
+
+    left_data_source_reference: DataSourceReference
+    right_data_source_reference: DataSourceReference
```

Comment:
I noticed this below, but do we need to have a different structure for multi-hop joins here, or are we just trying to give people the `identifier1__identifier2__target_data_source` marker?

In the future we'll definitely need something that's more explicit about the dependency sequencing, because we'll want to use the logic here to find data source join paths based on a source and target where the target is referred by dimension entity. Conceptually it'd look like `{entity: user, dimension: country}` and we'd be grabbing the path to that separately.